src/ovs/of_controller.cpp

```diff
@@ -56,26 +58,45 @@ void OFController::message_callback(OFConnection* ofconn, uint8_t type, void* da
 
             // setup default flows for each bridge
             if (bridge_name == "br-int") {
+                // ACA_LOG_DEBUG("%s\n", "Setting up default rules for br-int");
                 setup_default_br_int_flows();
             }
 
             if (bridge_name == "br-tun") {
+                // ACA_LOG_DEBUG("%s\n", "Setting up default rules for br-tun");
                 setup_default_br_tun_flows();
             }
         }
     } else if (type == fluid_msg::of13::OFPT_BARRIER_REPLY) {
         auto t = std::chrono::high_resolution_clock::now();
-        ACA_LOG_INFO("OFController::message_callback - recv OFPT_BARRIER_REPLY on %ld\n", t.time_since_epoch().count());
+        std::cout<<"OFController::message_callback - recv OFPT_BARRIER_REPLY on "<<t.time_since_epoch().count()<<std::endl;
+        // ACA_LOG_INFO("OFController::message_callback - recv OFPT_BARRIER_REPLY on %ld\n", t.time_since_epoch().count());
     } else if (type == fluid_msg::of13::OFPT_PACKET_IN) {
-        fluid_msg::of13::PacketIn *pin = new of13::PacketIn();
-        pin->unpack((uint8_t *) data);
-        uint32_t in_port = pin->match().in_port()->value();
-
-        // pass new allocated memory of packet-in to ACA_On_Demand_Engine to determine which type of request it is
-        aca_on_demand_engine::ACA_On_Demand_Engine::get_instance().thread_pool_.push(
-                std::bind(&aca_on_demand_engine::ACA_On_Demand_Engine::parse_packet,
-                          &aca_on_demand_engine::ACA_On_Demand_Engine::get_instance(),
-                          in_port, (void *)pin->data()));
+        packet_in_counter ++;
+        //// pass new allocated memory of packet-in to ACA_On_Demand_Engine to determine which type of request it is
+        //aca_on_demand_engine::ACA_On_Demand_Engine::get_instance().thread_pool_.push(
+        //        std::bind(&aca_on_demand_engine::ACA_On_Demand_Engine::parse_packet,
+        //                  &aca_on_demand_engine::ACA_On_Demand_Engine::get_instance(),
+        //                  in_port, (void *)pin->data(), ofconn->get_id()));
+        
+        // marl::schedule([=]{
+        //     std::string flow_string = "table=1,nw_dst=127.0.0.1,priority=1,actions=drop";
+        //     execute_flow(ofconn->get_id(), flow_string);
+        // });
+        
+        marl::schedule([=] {
+            fluid_msg::of13::PacketIn* pin = new fluid_msg::of13::PacketIn();
```

Comment:
there's a memory leak here, but I remember and will fix later

Reply:
Removed.

Reply:
Already fixed, by creating a copy of (*data) synchronously in message_call (before async job execution in marl), then at the end of message_callback function (*data) will be auto freed. 
Inside each marl task, the series of lines of code will be executed synchronously. So after pin is consumed in parse_packet() then we can safely free it.